### ISSUE TITLE
Pipes behave weirdly and inconsistently across platforms

### ISSUE THREAD

**Issue by maxried** [fish-shell/fish-shell#5177]
Please tell us which fish version you are using by executing the following:

1)
```
fish --version:
fish, version 2.7.1-1430-g56002abc-dirty
uname -a:
Darwin gotthold.fritz.box 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64
echo $TERM:
xterm-256color
```

2 - fish playground at https://rootnroll.com/d/fish-shell/ )
```
fish --version:
fish, version 2.7.1                                                                                         
uname -a:
Linux ac43aefa344b 4.14.11-coreos #1 SMP Fri Jan 5 11:00:14 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux         
echo $TERM:
xterm
```

3 - Android 9.0 September release inside of Termux )
```
fish --version:
fish, version 2.7.1                                                                                         
uname -a:
Linux localhost 4.4.116-g6d15b70 #1 PREEMPT Thu Aug 2 01:00:42 UTC 2018 aarch64 Android
echo $TERM:
xterm-256color
```

Problem:

I experience greatly inconsistent and weird behavior of pipes between three different platforms. Maybe I have a misunderstanding about the pipe concepts between different OSes and fish's implementation, so feel free to enlighten me. It feels to be related to buffering.

Platform 1 - macOS:
Good, works as I expect it:
```
max@gotthold ~$ sh -c 'env HOME=$(mktemp -d) fish'
Welcome to fish, the friendly interactive shell
max@gotthold /U/max> yes | grep y
y
y
y
y
y
y
y
[...]
```

No output whatsoever:
```
max@gotthold /U/max> yes | grep y | grep y
^C⏎
max@gotthold /U/max>
```

This one makes me believe it's related to buffering. These processes never terminate on their own and the output happens only when I press ^C:
```
max@gotthold /U/max> yes | grep y | head -n5
^Cy
y
y
y
y
max@gotthold /U/max>
```

In `htop` I can clearly see that there are only two process started while `yes | grep y | head -n5`. This seems to happen whenever the incoming stream won't terminate.

Platform 2 - Fish shell playground:
`yes | grep y | head -n5`: Nothing until you press ^C, then it explodes and writes 'y's. For a short moment it displays the following:
![image](https://user-images.githubusercontent.com/5188114/45298546-1d8e5380-b509-11e8-8ff2-c6c40e77e467.png)

`yes | grep y | grep y` behaves equivalently. No output until ^C, then plenty of 'y's.

`yes | grep y` works as expected.

Platform 3 - Android/Termux:
`yes | grep y`: Lots of 'y's
`yes | grep y | grep y`: Lots of 'y's
```
yes | grep y | head -n5:
y
y
y
y
```

Only the third platform works as I expect it to.

Why is this a problem? I have this real world example:
`adb logcat | grep Wifi | grep -v State`
`tailf /var/log/syslog | grep sgsdgsf | grep sfsdfsdfsbadfg`

**Comment by faho** [comment]
Duplicate of #1396. Fish buffers function output, and by default `grep` is a function.

This does not occur on Termux, because they don't seem to ship it (no "--color" option on theirs?).

**Comment by maxried** [comment]
Why is grep a function then if it causes inconsistencies? Could I have known this besides the issue you linked?

**Comment by ridiculousfish** [comment]
`grep` probably should not have a wrapper function, but the major architectural issue is the need to buffer function output. It's annoying for sure (though I think I've finally hit on the right way to fix it).

**Comment by floam** [comment]
On some systems, like my Mac, we create a wrapper function unnecessarily. We could get away with just setting environmental variables and not actually creating a new `grep` function in `fish.grep`. We would need to check grep's output to do this safely though. Newer versions of grep do not like for you to set `GREP_OPTIONS`.

**Comment by ridiculousfish** [comment]
This will just get fixed with the concurrency coming in 3.1.
